drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -180,40 +243,61 @@ static void adc_context_start_sampling(struct adc_context *ctx)
 {
 	struct adc_it8xxx2_data *data =
 		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+	uint32_t channels = ctx->sequence.channels;
+	uint8_t channel_count = 0;
+	int err;
 
 	data->repeat_buffer = data->buffer;
 
-	adc_enable_measurement(data->ch);
+	/*
+	 * The ADC sampling of it8xxx2 needs to read each channel
+	 * in sequence.
+	 */
+	while (channels) {
+		data->ch = find_lsb_set(channels) - 1;
+		channels &= ~BIT(data->ch);
+
+		adc_enable_measurement(data->ch);
+
+		channel_count++;
+	}
+
+	err = check_buffer_size(&ctx->sequence, channel_count);
+	if (err) {
+		return;
+	}
+
+	adc_context_on_sampling_done(&data->ctx, DEVICE_DT_INST_GET(0));
 }
 
 static int adc_it8xxx2_read(const struct device *dev,
 			    const struct adc_sequence *sequence)
 {
-	struct adc_it8xxx2_data *data = DEV_DATA(dev);
-	uint32_t channel_mask = sequence->channels;
-	uint8_t channel_count = 0;
-	int err = 0;
+	struct adc_it8xxx2_data *data = dev->data;
+	int err;
 
-	data->buffer = sequence->buffer;
-
-	while (channel_mask) {
-		adc_context_lock(&data->ctx, false, NULL);
-		data->ch = find_lsb_set(channel_mask) - 1;
+	adc_context_lock(&data->ctx, false, NULL);
+	err = adc_it8xxx2_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
 
-		err = adc_it8xxx2_start_read(dev, sequence);
-		if (err) {
-			return err;
-		}
+	return err;
+}
 
-		channel_mask &= ~BIT(data->ch);
-		channel_count++;
-		adc_context_release(&data->ctx, err);
-	}
+#if defined(CONFIG_ADC_ASYNC)
```

Comment:
`#ifdef`

Reply:
Can't this function be marked as "unused" so the warning does not happen but allows the function to be used in the table if CONFIG_ADC_ASYNC is defined and the optimizer will remove it if it is not defined?  If that is the case then this #if is not needed.

Reply:
@jackrosenthal 
Done.
@brockus-zephyr 
Because CONFIG_ADC_ASYNC is currently only used for testing, we can use #ifdef more appropriate.

Reply:
I am not sure how this is currently being used makes a difference. It will still only be included if CONFIG_ADC_ASYNC is defined.